build: Experimental multi-platform images

errordeveloper opened the pull request:
This PR enables multi-platform images (targeting `amd64` and `arm64` to start with). These will be built automatically built in GitHub Actions, and with Arm binaries all cross-compiled, and only packaging stages are executed using binfmt/qemu method.

Fixes: #9898 
Related: https://github.com/cilium/cilium/issues/11334

aanm commented on images/runtime/.dockerignore:
```diff
@@ -0,0 +1 @@
+Dockerfile
```
Maybe a sub directory of `contrib/packaging/docker` instead?

joestringer replied:
IMO `images/xxx/` is a bit easier to type, but my main concern here would be to organise all of the dockerfiles consistently under some path rather than spread across different locations.

errordeveloper replied:
Images are primary packaging format that is currently supported, it's mission-critical, so it deserves a top-level directory, which also make it easy to find.
A new contributor, or perhaps a user, who is wondering how are the Cilium images built, should be able to look in the repo, find `images/` directory and look no further.
Personally, when I see `contrib`, I don't look there as to me it stands for "collection of assorted bits that couldn't go anywhere else".

errordeveloper replied:
> organise all of the dockerfiles consistently under some path rather than spread across different locations.

Definitely one of the main goals for me, unfortunately cannot touch the existing ones just yet, but hopefully can get rid of some of them in 1.9.

aanm commented on images/runtime/Dockerfile:
```diff
@@ -0,0 +1,44 @@
+# SPDX-License-Identifier: Apache-2.0
+
+ARG COMPILERS_IMAGE=docker.io/cilium/image-compilers:818f48aa3eada1d23018f02f3c43a169c51eea1a@sha256:2f71fc0c9bdbb9a430ccb8cd86cb1fecc63acff464fc253fe3cd3748488e0e65
+ARG GOLANG_IMAGE=golang:1.14.4@sha256:d39a459086c75920390c9b9d42bbc7c52fef1fe7666e7633a02508c607c889d4
```
`docker.io/library/golang...`

errordeveloper replied:
Yes, something to update everywhere, including image-tools. Just leaving out as cosmetic thing for now.

aanm commented on images/runtime/Dockerfile:
```diff
@@ -0,0 +1,44 @@
+
+ARG COMPILERS_IMAGE=docker.io/cilium/image-compilers:818f48aa3eada1d23018f02f3c43a169c51eea1a@sha256:2f71fc0c9bdbb9a430ccb8cd86cb1fecc63acff464fc253fe3cd3748488e0e65
+ARG GOLANG_IMAGE=golang:1.14.4@sha256:d39a459086c75920390c9b9d42bbc7c52fef1fe7666e7633a02508c607c889d4
+ARG UBUNTU_IMAGE=ubuntu:20.04@sha256:8bce67040cd0ae39e0beb55bcb976a824d9966d2ac8d2e4bf6119b45505cee64
```
`docker.io/library/ubuntu...`

aanm commented on images/runtime/Dockerfile:
```diff
@@ -0,0 +1,44 @@
+COPY download-cni.sh /tmp/download-cni.sh
+RUN /tmp/download-cni.sh
+
+FROM --platform=linux/amd64 ${GOLANG_IMAGE} as go-builder
```
This should be `gops` instead of `go-builder`, right?

errordeveloper replied:
Well, it's a general go builder, gops is just one of the pieces that's being built by it (albeit only one at the moment).

aanm commented on images/builder/Dockerfile:
```diff
@@ -0,0 +1,23 @@
+#ARG CILIUM_RUNTIME_IMAGE=docker.io/errordeveloper/cilium-runtime:1d8b75fddd0ae3348c48392d12b6d6084e239c5b-dev-wip
```
leftover?

errordeveloper replied:
It's gone now, I believe.

aanm commented on images/builder/Dockerfile:
```diff
@@ -0,0 +1,23 @@
+#ARG CILIUM_RUNTIME_IMAGE=docker.io/errordeveloper/cilium-runtime:1d8b75fddd0ae3348c48392d12b6d6084e239c5b-dev-wip
+ARG COMPILERS_IMAGE=docker.io/cilium/image-compilers:818f48aa3eada1d23018f02f3c43a169c51eea1a@sha256:2f71fc0c9bdbb9a430ccb8cd86cb1fecc63acff464fc253fe3cd3748488e0e65
+ARG GOLANG_IMAGE=golang:1.14.4@sha256:d39a459086c75920390c9b9d42bbc7c52fef1fe7666e7633a02508c607c889d4
```
`docker.io/library/golang...`

errordeveloper replied:
Fixed here, also in https://github.com/cilium/image-tools/pull/42.

aanm commented on images/builder/Dockerfile:
```diff
@@ -0,0 +1,23 @@
+ARG COMPILERS_IMAGE=docker.io/cilium/image-compilers:818f48aa3eada1d23018f02f3c43a169c51eea1a@sha256:2f71fc0c9bdbb9a430ccb8cd86cb1fecc63acff464fc253fe3cd3748488e0e65
+ARG GOLANG_IMAGE=golang:1.14.4@sha256:d39a459086c75920390c9b9d42bbc7c52fef1fe7666e7633a02508c607c889d4
+
+#FROM ${COMPILERS_IMAGE} as compilers-dist
```
leftover?

errordeveloper replied:
Yes, I think I already pushed a commit that deletes this.

aanm commented on images/builder/Dockerfile:
```diff
@@ -0,0 +1,23 @@
+#FROM ${COMPILERS_IMAGE} as compilers-dist
+FROM ${GOLANG_IMAGE} as golang-dist
+
+#FROM ${CILIUM_RUNTIME_IMAGE}
```
leftover?

aanm commented on images/builder/Dockerfile:
```diff
@@ -0,0 +1,23 @@
+FROM ${COMPILERS_IMAGE}
+LABEL maintainer="maintainer@cilium.io"
+
+#COPY --from=compilers-dist / /
```
leftover?

aanm commented on images/builder/install-builder-deps.sh:
```diff
@@ -0,0 +1,28 @@
+
+apt-get update
+
+ln -fs /usr/share/zoneinfo/UTC /etc/localtime
```
what's this for?

errordeveloper replied:
For some reason new version of Ubuntu requires user to set timezone for some packages, I don't know why and for which packages exactly it was...

errordeveloper replied:
My guess is that from Ubuntu distro point of view, there is no such thing as default timezone any more.

aanm replied:
This was a clue to leave a comment in script describing what is it for.

errordeveloper replied:
I've added a comment above, it actually got me puzzled when I first noticed it in a Dockerfile that Daniel wrote.

aanm commented on images/runtime/Dockerfile:
```diff
@@ -35,6 +35,7 @@ COPY iptables-wrapper /usr/sbin/iptables-wrapper
 COPY configure-iptables-wrapper.sh /tmp/configure-iptables-wrapper.sh
 RUN /tmp/configure-iptables-wrapper.sh
 
+# XXX: perhaps these should be copied into /usr/local/bin, so that it's easy to copy-out
```
leftover? I think copied into `/usr/local/bin` makes sense if it's not too much work to do it so.

errordeveloper replied:
Yeah, I think so, also in each of the upstream images.

aanm commented on images/builder/Dockerfile:
```diff
@@ -1,16 +1,11 @@
-#ARG CILIUM_RUNTIME_IMAGE=docker.io/errordeveloper/cilium-runtime:1d8b75fddd0ae3348c48392d12b6d6084e239c5b-dev-wip
```
This commit can be squashed with a previous one

aanm commented on images/cilium/Dockerfile:
```diff
@@ -0,0 +1,35 @@
+# syntax=docker/dockerfile:1.1-experimental
+
+#FROM quay.io/cilium/hubble:v0.6.0 as hubble
+
```
Leftover?

aanm commented on images/cilium/Dockerfile:
```diff
@@ -0,0 +1,35 @@
+
+#FROM quay.io/cilium/hubble:v0.6.0 as hubble
+
+FROM --platform=linux/amd64 docker.io/errordeveloper/cilium-builder:019538f3bd3275e8d8369e2c5823efbe0f3a4f06-dev as builder
```
what's the process for us to replace `errordeveloper/cilium-builder` with `cilium/cilium-builder`?

errordeveloper replied:
Still a WIP PR, please ignore this. It's a topic to discuss.

aanm commented on images/cilium/Dockerfile:
```diff
@@ -0,0 +1,35 @@
+#COPY --from=hubble /usr/bin/hubble /usr/bin/hubble
+#COPY --from=builder /go/src/github.com/cilium/cilium/plugins/cilium-cni/cni-install.sh /cni-install.sh
+#COPY --from=builder /go/src/github.com/cilium/cilium/plugins/cilium-cni/cni-uninstall.sh /cni-uninstall.sh
+#COPY --from=builder /go/src/github.com/cilium/cilium/contrib/packaging/docker/init-container.sh /init-container.sh
```
This needs to go in.

errordeveloper replied:
Yes, it's still a WIP PR.

aanm commented on images/cilium/Dockerfile:
```diff
@@ -17,9 +19,10 @@ RUN --mount=type=bind,readwrite,target=/go/src/github.com/cilium/cilium --mount=
-LABEL maintainer="maintainer@cilium.io"
+FROM ${CILIUM_RUNTIME_IMAGE}
 ARG TARGETPLATFORM
+LABEL maintainer="maintainer@cilium.io"
```
I think this is also missing the `ARG CILIUM_SHA=""`, right?

errordeveloper replied:
That was only needed due to how image GC worked on Jenkins.

joestringer replied:
To clarify, do we want the Cilium SHA as a label for subsequent introspection?

errordeveloper replied:
It's actually being used as image tag for the agent image, as well `operator` and `hubble-relay`.

aanm commented on images/cilium/Dockerfile:
```diff
@@ -10,14 +10,14 @@ ARG NOSTRIP
 ARG LOCKDEBUG
```
What happened to the `ARG V`?

joestringer replied:
Context: `ARG V` allows developers to specify a verbosity flag where for instance I define it to `V=0` in my shell to cause the build system to be less verbose. This allows devs to get a clearer picture of what fails, while still printing any and all possibly relevant information during builds on automated systems where you may need to introspect later.

errordeveloper replied:
I agree that control of verbosity can useful, but GitHub Actions hides logs for each step, so there is no reason to worry about build log verbosity.
Secondly, when building locally BuildKit only presents per-step logs on failure, which is very different from old `docker build` output format.
Personally, I don't think there is a better way than passing `V` from make to docker, and back to make, so I dropped `V` for now, as this PR is focus only on enabling builds in GitHub Actions. I'd rather provide high-level flags as part of a higher-level tool (cilium/image-tools#10), which would also enable a way of building all of the images more consistently and avoid the boilerplate. At least that is my line of thinking at present, all that is up for a discussion, of course.

joestringer replied:
I think we can evaluate how this affects developer use of these build targets when we start migrating developers over to this. Right now this has not been a target and the current developer use cases are not intended to be superseded by this system at least in the short term.

aanm commented on images/cilium/Dockerfile:
```diff
@@ -10,14 +10,14 @@ ARG NOSTRIP
 ARG LOCKDEBUG
```
commit can be squashed

errordeveloper replied:
I'll probably squash the whole lot at the end into one or two, still WIP.

aanm reviewed (dismissed):
looks good overall. I noticed that all dockerfiles are building both architectures sequentially. Is it possible to only build amd64 or only build arm64? I'm thinking in the case that a developer that only wants to build an amd64 image does not have to wait for arm64 to build for example.

joestringer commented on images/runtime/download-cni.sh:
```diff
@@ -0,0 +1,20 @@
+set -o pipefail
+set -o nounset
+
+cni_version="0.7.5"
```
Might be nice to factor all of these version defines into a separate shell script purely for versions and import that into each of these shell scripts. That way we have exactly one place to check for all of our versions and we can iterate through & evaluate whether to bump them.

joestringer replied:
(I see at least ineffassign, gops and cni here so far, I'm guessing there'll be more)

errordeveloper replied:
Yeah, definitely something to think about, for now I just make sure versions are set as variables at the top (and `ARGS` in Dockerfile).

The current take overall is that each of the images that are self-contained to a single directory is defined by that directory, hence the use of tree hashes for image versioning.

I am actually thinking it'd be nice to have a YAML file with versions and other build prams set out, and allow developers to override it locally for testing.

joestringer commented on images/scripts/build-image.sh:
```diff
@@ -0,0 +1,105 @@
+MAKER_IMAGE="${MAKER_IMAGE:-docker.io/cilium/image-maker:bc81755ec8f6c5afcb10a416cef73f99a35fee2c}"
+
+if [ "$#" -lt 6 ] ; then
+  echo "$0 supports minimum 6 argument"
```
nit:

```suggestion
  echo "usage: $0 IMAGE_NAME IMAGE_DIR PLATFORM OUTPUT_DIR BUILDER REGISTRY [REGISTRY ...]"
```

joestringer replied:
also nit, >&2 to go to stderr

joestringer commented on images/scripts/make-image-tag.sh:
```diff
@@ -0,0 +1,68 @@
+# `git name-rev --name-only --tags --no-undefined HEAD` however Cilium usually
+# gets two tags one with `v` prefix and one without, beacause of that the output
+# of `git name-rev` maybe one or the other; to implement that with certainty
+# release process/automation needs to be reviewed in detail
```
There's some background here where IIRC readthedocs doesn't recognize `vX.Y.Z` tags so we have to tag `X.Y.Z` to get RTD to generate the "stable" docs based on a newly released version; at the same time IIRC Github encourages `vX.Y.Z` style tags via its releases page mechanisms, I don't recall the details.

One potential mitigation may be moving docs build into our own infrastructure and just pushing static pages to a simpler, dumber webserver. But that's in general a bigger question.

joestringer replied:
(I don't think we ever tried to get RTD to recognize `vX.Y.Z`, maybe that's another path to explore)

errordeveloper replied:
I actually fixed this TODO, will push new version soon.

joestringer commented on images/scripts/make-image-tag.sh:
```diff
@@ -0,0 +1,68 @@
+# release process/automation needs to be reviewed in detail
+
+if [ "$#" -gt 1 ] ; then
+  echo "$0 supports exactly 1 or no arguments"
```
```suggestion
  echo "usage: $0 [IMAGE_DIR]"
```

errordeveloper replied:
Maybe, I'm just keeping these simple for now. In the future, I'd rather turn these into a Go program actually (see https://github.com/cilium/image-tools/issues/10). I think that way we could enforce best practice and avoid all the boilerplate and manual steps.

joestringer commented on images/operator/Dockerfile:
```diff
@@ -0,0 +1,29 @@
+COPY --from=builder /out/${TARGETPLATFORM} /
+
+WORKDIR /
+CMD ["/usr/bin/cilium-operator-generic"]
```
what's the relationship between this dockerfile and the current per-provider docker image building?

errordeveloper replied:
To be honest, I'm just trying to keep it simple here, this image gets all 4 binaries, but I'd like to find a way to have separate images without having separate (largely duplicate) Dockerfiles per-se. Part of me is really keen to explore abstracting the Dockerfiles into some simple YAML format and generating Dockerfile on the fly, or using syntax extension mechanism that [BuildKit now offers](https://matt-rickard.com/building-a-new-dockerfile-frontend/).

joestringer commented on images/cilium/Dockerfile:
```diff
@@ -0,0 +1,38 @@
+#COPY --from=builder /go/src/github.com/cilium/cilium/plugins/cilium-cni/cni-install.sh /cni-install.sh
+#COPY --from=builder /go/src/github.com/cilium/cilium/plugins/cilium-cni/cni-uninstall.sh /cni-uninstall.sh
+#COPY --from=builder /go/src/github.com/cilium/cilium/contrib/packaging/docker/init-container.sh /init-container.sh
+
```
Where did `cilium/cilium-envoy` end up?

errordeveloper replied:
It requires more work, I've managed to compile Envoy itself with a few work-arounds, but making it happen for proxy is going to take a few days. From discussion with @aanm It seems reasonable to omit L7 support from the first version.

joestringer replied:
I think that's fine as long as we're clear about the limitations and any workarounds necessary to make that work. For instance, I believe that on startup Cilium will check the envoy binary availability and version to ensure it's compatible. I'm pretty sure there's a flag to override this, but what's the flag? etc.

joestringer reviewed (commented):
Since it's a draft, I didn't do a thorough pick through but I pointed out a few first-round things below.

errordeveloper committed:
build: Import key scripts from cilium/image-tools@05025951392aaff17a5828b8052c2535f079667f

This is a temporary measure, these scripts should be consumed either as a submodule,
or a portable tool (see cilium/image-tools#10).

Signed-off-by: Ilya Dmitrichenko <errordeveloper@gmail.com>

errordeveloper committed:
build: Disable hadolint, adjust subdirs

hadolint doesn't yet support experimental `RUN --mount` syntax.

Signed-off-by: Ilya Dmitrichenko <errordeveloper@gmail.com>

errordeveloper committed:
build: New runtime image with multi-platform support

Signed-off-by: Ilya Dmitrichenko <errordeveloper@gmail.com>

errordeveloper committed:
build: New builder image supporting cross-compilation

Signed-off-by: Ilya Dmitrichenko <errordeveloper@gmail.com>

errordeveloper committed:
build: Add new multi-platform cilium image

This image ommits Hubble CLI and Envoy for now, as neither
have arm64 builds at the moment.

Signed-off-by: Ilya Dmitrichenko <errordeveloper@gmail.com>

errordeveloper committed:
build: Add new multi-platform operator image

This intial version includes all operator falavours, since
parametrisation needs some more work to be fully integrated,
and it's better to avoid duplication of Dockerfiles.

Signed-off-by: Ilya Dmitrichenko <errordeveloper@gmail.com>

errordeveloper committed:
build: Add new multi-platform hubble-relay image

Signed-off-by: Ilya Dmitrichenko <errordeveloper@gmail.com>

errordeveloper committed:
build: Add Makefile and GitHub Actions workflow for new images

Signed-off-by: Ilya Dmitrichenko <errordeveloper@gmail.com>

errordeveloper committed:
build: Add helper scripts to update runtime and builder images

Signed-off-by: Ilya Dmitrichenko <errordeveloper@gmail.com>

joestringer reviewed (approved):
Looks like a reasonable start point. As per my prior comment, I don't think it's a great idea to have this in v1.8 just because it is subject to further changes and we'll figure out what works / what needs improvement over time.

errordeveloper committed:
CODEOWNERS: Add GitHub Actions workflows and new images dir to @cilium/build group

Signed-off-by: Ilya Dmitrichenko <errordeveloper@gmail.com>

errordeveloper commented on images/builder/Dockerfile:
```diff
@@ -0,0 +1,18 @@
+ARG COMPILERS_IMAGE=docker.io/cilium/image-compilers:818f48aa3eada1d23018f02f3c43a169c51eea1a@sha256:2f71fc0c9bdbb9a430ccb8cd86cb1fecc63acff464fc253fe3cd3748488e0e65
```
Missing license header

errordeveloper commented on images/cilium/Dockerfile:
```diff
@@ -0,0 +1,41 @@
+# syntax=docker/dockerfile:1.1-experimental
```
Missing license header

errordeveloper commented on images/cilium/Dockerfile.dockerignore:
```diff
@@ -0,0 +1,28 @@
+# since this is image is built with root context, most of the files
```
Missing license header

errordeveloper commented on images/hubble-relay/Dockerfile:
```diff
@@ -0,0 +1,29 @@
+# syntax=docker/dockerfile:1.1-experimental
```
Missing license header

errordeveloper commented on images/hubble-relay/Dockerfile.dockerignore:
```diff
@@ -0,0 +1,20 @@
+# since this is image is built with root context, most of the files
```
Missing license header

errordeveloper commented on images/operator/Dockerfile:
```diff
@@ -0,0 +1,29 @@
+# syntax=docker/dockerfile:1.1-experimental
```
Missing license header

errordeveloper commented on images/operator/Dockerfile.dockerignore:
```diff
@@ -0,0 +1,20 @@
+# since this is image is built with root context, most of the files
```
Missing license header

errordeveloper committed:
build: Add documentation for next-gen images

Signed-off-by: Ilya Dmitrichenko <errordeveloper@gmail.com>

errordeveloper commented on images/cilium/Dockerfile:
```diff
@@ -0,0 +1,45 @@
+
+RUN --mount=type=bind,readwrite,target=/go/src/github.com/cilium/cilium --mount=target=/root/.cache,type=cache --mount=target=/go/pkg/mod,type=cache \
+  env GOARCH=arm64 CC=aarch64-linux-gnu-gcc \
+    make clean-container build-container install-container \
```
Turns out there is something odd going on and `/var/lib/cilium` is not populated properly.

errordeveloper replied:
This is due to exclusion of `.git` dir, waiting for https://github.com/cilium/cilium/pull/12326.

errordeveloper replied:
#12326 was merged, and I've rebased this PR just now

aanm commented on images/cilium/Dockerfile:
```diff
@@ -0,0 +1,45 @@
+  make clean-container build-container install-container \
+    NOSTRIP=$NOSTRIP LOCKDEBUG=$LOCKDEBUG PKG_BUILD=1 SKIP_DOCS=true DESTDIR=/out/linux/amd64
+
+RUN --mount=type=bind,readwrite,target=/go/src/github.com/cilium/cilium --mount=target=/root/.cache,type=cache --mount=target=/go/pkg/mod,type=cache \
+  env GOARCH=arm64 CC=aarch64-linux-gnu-gcc \
+    make clean-container build-container install-container \
+      NOSTRIP=$NOSTRIP LOCKDEBUG=$LOCKDEBUG PKG_BUILD=1 SKIP_DOCS=true DESTDIR=/out/linux/arm64 \
+      HOST_CC=aarch64-linux-gnu-gcc HOST_STRIP=aarch64-linux-gnu-strip
```
I don't understand why we are duplicating these lines. To me this looks that this could be :
```
ARG GOARCH
ARG ARCH
ARG OS
RUN --mount=type=bind,readwrite,target=/go/src/github.com/cilium/cilium --mount=target=/root/.cache,type=cache --mount=target=/go/pkg/mod,type=cache \
  env GOARCH=$GOARCH CC=$ARCH-$OS-gnu-gcc \
    make clean-container build-container install-container \
      NOSTRIP=$NOSTRIP LOCKDEBUG=$LOCKDEBUG PKG_BUILD=1 SKIP_DOCS=true DESTDIR=/out/$OS/$GOARCH \
      HOST_CC=$ARCH-$OS-gnu-gcc HOST_STRIP=$ARCH-$OS-gnu-strip
```

which works for both `arm64` and `amd64` so we don't have to waste resources building 2 architectures if, in the end we will only copy one of them (line 38 `COPY --from=builder /out/${TARGETPLATFORM} /`)

errordeveloper replied:
>  in the end we will only copy one of them 

The way to understand it is that there is an implicit loop and `TARGETPLATFORM` is the loop variable.

The build runs only for `amd64`, as it has `FROM --platform=linux/amd64` (line 9).
But there is no `--platfrom` flag on line 34, so that stages runs on each of the platforms.

aanm replied:
Also, why are we mounting `/go/pkg/mod`? Cilium vendors all code in the `vendor` directory

aanm replied:
> > in the end we will only copy one of them
> 
> The way to understand it is that there is an implicit loop and `TARGETPLATFORM` is the loop variable.
> 

But from line 9 to line 33 we are building for both archs at the same time, no?

errordeveloper replied:
> But from line 9 to line 33 we are building for both archs at the same time, no?

No, that's from line 34 onwards.

> Also, why are we mounting `/go/pkg/mod`? Cilium vendors all code in the `vendor` directory

That is just a matter of re-using a pattern. Do note that this directory is mounted as cache, so if something does end-up being written there it will get cached, but if nothing - nothing will get cached.

aanm reviewed (changes requested):
The duplication of builds that I've point it out in the Cilium Dockerfile seems to be done for all dockerfiles. I think this can be avoided.

aanm commented on images/scripts/lint.sh:
```diff
@@ -12,8 +12,7 @@ MAKER_IMAGE="${MAKER_IMAGE:-docker.io/cilium/image-maker:bc81755ec8f6c5afcb10a41
 
 if [ -z "${MAKER_CONTAINER+x}" ] ; then
-   exec docker run --rm --volume "${root_dir}:/src" --workdir /src "${MAKER_IMAGE}" "/src/scripts/$(basename "${0}")"
+   exec docker run --rm --volume "${root_dir}:/src" --workdir /src/images "${MAKER_IMAGE}" "/src/images/scripts/$(basename "${0}")"
```
Can be squashed with the previous commit

errordeveloper replied:
The thing is that previous commit copies a specific revision of this script from `cilium/image-tools`, and this one commit represents a local tweak.

aanm commented on images/runtime/build-gops.sh:
```diff
@@ -0,0 +1,23 @@
+unset GOPATH
+
+go mod init github.com/cilium/cilium/images/runtime
+go get github.com/google/gops@v0.3.6
```
This should be `v0.3.10` https://github.com/cilium/cilium/blob/7e885d92abdcbd17254228ab6a25d40211c9d52e/contrib/packaging/docker/Dockerfile.runtime#L35

aanm commented on images/runtime/download-cni.sh:
```diff
@@ -0,0 +1,20 @@
+cni_version="0.7.5"
+
+for arch in amd64 arm64 ; do
+  curl --fail --show-error --silent --location "https://github.com/containernetworking/plugins/releases/download/v${cni_version}/cni-plugins-${arch}-v${cni_version}.tgz" --output "/tmp/cni-${arch}.tgz"
```
please verify the the contents of the `tgz`, similar to what is being done in https://github.com/cilium/cilium/blob/7e885d92abdcbd17254228ab6a25d40211c9d52e/contrib/packaging/docker/Dockerfile.runtime#L50-L53

errordeveloper replied:
Thanks, I've added these, but instead of downloading the checksum at runtime I add a file where checksums are declared, since the checksum file itself can be compromised.

aanm commented on images/runtime/download-cni.sh:
```diff
@@ -0,0 +1,20 @@
+set -o pipefail
+set -o nounset
+
+cni_version="0.7.5"
```
should be `0.8.6`

aanm commented on images/builder/Dockerfile:
```diff
@@ -0,0 +1,34 @@
+COPY install-protoc.sh /tmp/install-protoc.sh
+RUN /tmp/install-protoc.sh
+
+FROM ${TESTER_IMAGE} as test
+COPY --from=rootfs / /
+COPY test /test
+RUN /test/bin/cst
```
Would be nice to have comments stating what is this `cst` binary

errordeveloper replied:
This is meant to be documented in [`images/READE.md`](https://github.com/cilium/cilium/blob/1288388130a37625487477e4a95f42b031e5b169/images/README.md#testing).

aanm commented on images/cilium/Dockerfile:
```diff
@@ -0,0 +1,45 @@
+ARG CILIUM_BUILDER_IMAGE=docker.io/cilium/cilium-builder-dev:a116c9e46733fce354b871dd6e88e037d9459286
+ARG CILIUM_RUNTIME_IMAGE=docker.io/cilium/cilium-runtime-dev:bf5a13ab0dd341b446ec1f40a809d365024eeb09
+
+FROM --platform=linux/amd64 ${CILIUM_BUILDER_IMAGE} as builder
```
Please leave a comment to what this `--platform` is doing.

errordeveloper replied:
I've documented a simplified version of this pattern in the README, since it applied to other Dockerfiles also.

aanm commented on .github/workflows/images.yaml:
```diff
@@ -0,0 +1,67 @@
+          DOCKER_HUB_USERNAME: ${{ secrets.DOCKER_HUB_USERNAME }}
+        with:
+          entrypoint: make
+          args: -C images builder-image PUSH=true
```
Why do we need to build the runtime and builder images each time? We shouldn't build those as part of the release process because we need to build these first, test them against the CI and only them publish the Cilium images based on these.

Perhaps we should have another GH action that will build these images independently?

This comment can be addressed as a follow up.

errordeveloper replied:
Actually, these won't be rebuilt unless there were changes. runtime and builder are scoped to their subdirs, and we use git tree hashes as tags. The logic is implemented in [`image/scripts/build-image.sh`](https://github.com/cilium/cilium/blob/ea59fd095d7c38f70a82a9498219f69f1f6a192c/images/scripts/build-image.sh#L72-L92). So with runtime and builder the tag changes only when there are actual changes in the directory.

aanm replied:
Please add a comment here about that behavior.

errordeveloper replied:
Sure, I added a comment here. There is a mention in the readme also.

aanm commented on .github/workflows/images.yaml:
```diff
@@ -0,0 +1,67 @@
+          entrypoint: make
+          args: -C images cilium-image PUSH=true
+      - uses: docker://docker.io/cilium/image-maker:bc81755ec8f6c5afcb10a416cef73f99a35fee2c
+        name: Run make operator-image
```
`s/operator-image/operator-images/`

errordeveloper replied:
It's only one image here at the moment. I'd like to have a better story for handling image variants, so keeping it simple for the time being.

aanm replied:
So we are not building `operator-generic`, `operator-aws`, `operator-azure`?

errordeveloper replied:
We are building all the binaries, but those end-up inside of the one image. As I said, just a thing for now and I'll open follow-up issues.

aanm commented on images/Makefile:
```diff
@@ -25,9 +25,15 @@ lint:
 runtime-image: .buildx_builder
 	TEST=true scripts/build-image.sh cilium-runtime-dev images/runtime linux/amd64,linux/arm64 $(OUTPUT) "$$(cat .buildx_builder)" $(REGISTRIES)
 
+update-runtime-image:
```
Please add more info in the commit, it's not clear to me what is the purpose of these scripts / makefile targets.

errordeveloper replied:
This is meant to be documented in [`images/READE.md`](https://github.com/cilium/cilium/blob/ea59fd095d7c38f70a82a9498219f69f1f6a192c/images/README.md), have you had a look at it yet?

nebril commented on images/README.md:
```diff
@@ -0,0 +1,96 @@
+# Experimental Nex-gen Cilium Images
```
nit: s/Nex/Next

errordeveloper replied:
Thank you!

nebril commented on images/runtime/download-cni.sh:
```diff
@@ -0,0 +1,21 @@
+source /tmp/cni-version.sh
+
+for arch in amd64 arm64 ; do
+  curl --fail --show-error --silent --location "https://github.com/containernetworking/plugins/releases/download/v${cni_version}/cni-plugins-linux-${arch}-v${cni_version}.tgz" --output "/tmp/cni-${arch}.tgz"
```
Should we add retries here in case there is a temporary connectivity failure?

errordeveloper replied:
Perhaps, but I'd rather leave that for another PR, and ideally we should update all calls to curl in `cilium/image-tools` too.

aanm commented on images/runtime/build-gops.sh:
```diff
@@ -0,0 +1,26 @@
+unset GOPATH
+
+# when updating this version, also update contrib/packaging/docker/Dockerfile.runtime
+gobs_version="0.3.10"
```
`s/gobs/gops/`

aanm commented on images/runtime/build-gops.sh:
```diff
@@ -0,0 +1,26 @@
+gobs_version="0.3.10"
+
+go mod init github.com/cilium/cilium/images/runtime
+go get "github.com/google/gops@v${gobs_version}"
```
`s/gobs/gops/`
